Builtin functions for linear regression algorithms

kev-inn opened the pull request:
This adds 4 new DML-bodied builtin functions:

  * lm
  * lmDS
  * lmCG
  * lmpredict

These functions are to be used for linear regression.
`lm` calls either `lmDS` or `lmCG` depending on the size of the input matrices. `lmpredict` can be used
to predict the class of a feature vector with the weights usually calculated before by use of `lm`.

This also adds a file used for documentation of builtin functions, which currently only documents those new added builtins.

Beware that lm currently is non-functional and the test cases will fail, because of a namespace conflict when using multiple different DML-bodied builtin functions.

kev-inn committed:
[SYSTEMDS-112] Builtin functions for linear regression algorithms

This adds 4 new DML-bodied builtin functions:

  * lm
  * lmDS
  * lmCG
  * lmpredict

These functions are to be used for linear regression.
`lm` calls either `lmDS` or `lmCG` depending on the size of the input matrices. `lmpredict` can be used
to predict the class of a feature vector with the weights usually calculated before by use of `lm`.

This also adds a file used for documentation of builtin functions, which currently only documents those new added builtins.

Beware that lm currently is non-functional and the test cases will fail, because of a namespace conflict when using multiple different DML-bodied builtin functions.

kev-inn committed:
[SYSTEMDS-112] Builtin functions for linear regression algorithms

This adds 4 new DML-bodied builtin functions:

  * lm
  * lmDS
  * lmCG
  * lmpredict

These functions are to be used for linear regression.
`lm` calls either `lmDS` or `lmCG` depending on the size of the input matrices. `lmpredict` can be used
to predict the class of a feature vector with the weights usually calculated before by use of `lm`.

This also adds a file used for documentation of builtin functions, which currently only documents those new added builtins.

Beware that lm currently is non-functional and the test cases will fail, because of a namespace conflict when using multiple different DML-bodied builtin functions.

kev-inn committed:
Adapt default reg and tol values to R

kev-inn committed:
Merge branch 'master' into linear_regression

mboehm7 committed:
[SYSTEMDS-122] Reduced sparse buffer pool serialization overhead

This patch makes the shallow serialize threshold slightly more
aggressive, which means that we accept larger in-memory overhead in
order to avoid deep serialization of sparse matrices in the buffer pool.

On a parfor data augmentation application with 10K input images and 20K
output images, this patch improves performance from 9.2s (139s left
indexing, 46s buffer pool release) to 6.4s (82s left indexing, 7.9s
buffer pool release).

mboehm7 commented on docs/builtins-reference.md:
```diff
@@ -0,0 +1,176 @@
+<!--
+{% comment %}
+Modifications Copyright 2018 Graz University of Technology
```
Please use ... 2019.

mboehm7 commented on docs/builtins-reference.md:
```diff
@@ -0,0 +1,176 @@
+    
+# Introduction
+
+Declarative Machine Learning has builtin functions which enable access to both low- and high-level functions
```
Maybe "The DML (...) language has ..."

mboehm7 commented on docs/builtins-reference.md:
````diff
@@ -0,0 +1,176 @@
+### Example
+```r
+X = rand (rows = 50, cols = 10)
+y = rand (rows = 50, cols = 1)
````
Please make it more meaningful by using `y = X %*% rand(rows=ncol(X), 1)` (for all examples)

mboehm7 commented on scripts/builtin/lm.dml:
```diff
@@ -0,0 +1,28 @@
+
+m_lm = function(Matrix[Double] X, Matrix[Double] y, Integer icpt = 0, Double reg = 1e-7, Double tol = 1e-7, Integer maxi = 0, Boolean verbose = TRUE)
+    return (Matrix[Double] B) {
+  n = nrow (X);
```
Please inline the nrow(X) as it is used just once

mboehm7 commented on scripts/builtin/lm.dml:
```diff
@@ -0,0 +1,28 @@
+  }
+  else {
+    B = lmCG(X, y, icpt, reg, tol, maxi, verbose)
+  }
```
Remove the unnecessary curly braces for these one-line conditions

mboehm7 commented on scripts/builtin/lmCG.dml:
```diff
@@ -0,0 +1,180 @@
+#-------------------------------------------------------------
+#
+# Copyright 2019 Graz University of Technology
```
Use the modifications-license as the original linregCG was copied from Apache SystemML.

kev-inn replied:
I am sorry could you explain which comment I would have to add instead. Is it the same that as the one inside LinearRegCG.dml?

kev-inn replied:
Found one in DenseBlockDRB, will add it

mboehm7 replied:
exactly

mboehm7 commented on scripts/builtin/lmDS.dml:
```diff
@@ -0,0 +1,139 @@
+#-------------------------------------------------------------
+#
+# Copyright 2019 Graz University of Technology
```
Use the modifications-license as the original linregDS was copied from Apache SystemML.

mboehm7 commented on src/test/scripts/functions/builtin/lm.R:
```diff
@@ -0,0 +1,26 @@
+#-------------------------------------------------------------
+#
+# Copyright 2018 Graz University of Technology
```
2019

mboehm7 commented on scripts/builtin/lmpredict.dml:
```diff
@@ -0,0 +1,21 @@
+#-------------------------------------------------------------
+
+m_lmpredict = function(Matrix[Double] X, Matrix[Double] w) return (Matrix[Double] y) {
+  y = X %*% w
```
Please extend this to allow for specify intercept as well - in which case you would need to append a column of ones and/or scale / shift.

mboehm7 replied:
For now it's sufficient to support intercept 1 and include a TODO for intercept 2 as this is a longer discussion

mboehm7 reviewed (commented):
Thanks @kev-inn - this is a good start. Please incorporate the comments and then I'll merge it in.

mboehm7 committed:
[MINOR] Various fixes of testcases (num jobs, ext functions)

mboehm7 committed:
[SYSTEMDS-11] Fix namespace handling of dml-bodied builtin functions

This patch fixes issues with using multiple dml-bodied builtin functions
in the same script. So far the default namespace was bound to the first
dml-bodied function file path. Now, we properly handle the file paths
for the default namespace separately.

mboehm7 commented:
@kev-inn Thanks again for catching the issue regarding namespace handling of multiple dml-bodied builtin functions. This issue has been fixed w/ commit 506cc9bae9a6a6135233d85cd84825a3963bb936.

corepointer committed:
[SYSTEMDS-112,122] Additional parfor image data augmentation tests

parfor test on image operations mirror, brightness using synthetic data

Closes #15.

kev-inn commented:
> @kev-inn Thanks again for catching the issue regarding namespace handling of multiple dml-bodied builtin functions. This issue has been fixed w/ commit [506cc9b](https://github.com/tugraz-isds/systemds/commit/506cc9bae9a6a6135233d85cd84825a3963bb936).

@mboehm7 I think I detected another case where we get problems now. Calling a builtin in a dml-bodied bultin leads to the error `namespace ./scripts/builtin is not defined`. You can test this with lm once I resolved all the issues.

mboehm7 commented:
ok thanks for catching - then let's finalize this PR and once it's in master, I'll have a look into this issue to avoid redundancy.

kev-inn committed:
Resolve PR problem

mboehm7 commented:
LGTM - thanks @kev-inn for incorporating the comments. I think that is an excellent start toward higher-level abstractions in SystemDS. During the merge I made only minor formatting changes and I moved the stats computation of lmCG and lmDS in a verbose branch - dead code elimination would have avoided unnecessary operations as well but this makes it more explicit (we can better understand the performance implications on script level).
